modules/end-user-guide/pages/url-parameter-for-configuring-project-remotes.adoc

```diff
@@ -0,0 +1,49 @@
+:_content-type: CONCEPT
+:description: URL parameter for configuring project remotes
+:keywords: configuring-remotes, configuring-remotes-parameter, remotes-parameter, project-remotes-parameter
+:navtitle: URL parameter for configuring project remotes
+:page-aliases:
+
+[id="url-parameter-for-configuring-project-remotes"]
+= URL parameter for configuring project remotes
+
+When you visit a URL for starting a new workspace, one Git remote will be configured by default. Git remotes can be configured manually in the command line after workspace startup, or manually by editing the workspace's {devworkspace} {orch-name} resource after workspace creation. 
+
+To have {prod-short} automatically configure Git remotes on workspace creation, the `remotes=<remotes>` URL parameter can be used.
+
+The first remote specified in the `remotes` parameter will have it's default branch checked out on workspace startup.
+
+The value of the `remotes` parameter can be provided in two formats: by providing both remote names and URLs, and by providing only URLs.
+
+== Providing remote names and URLs
+
+Remote names and URLs can be specified by providing a list of `{name,URL}` pairs in the following format:
+
+[source,subs="+quotes,+attributes,+macros"]
+----
+{{origin,https://git-url/origin/project.git},{upstream,https://git-url/upstream/project.git}, ... } <1>
+----
+
+The following example configures two remotes on workspace creation, `origin` and `upstream`. The `origin` remote's default branch is checked out on workspace startup.
+[source,subs="+quotes,+attributes,+macros"]
+----
+pass:c,a,q[{prod-url}]#__<git_repository_url>__?remotes={{origin,https://git-url/origin/project.git},{upstream,https://git-url/upstream/project.git}}
+----
+
+== Providing only URLs
+
+Remote URLs can be specified by providing a list of URLs in the following format:
+
+[source,subs="+quotes,+attributes,+macros"]
+----
+{https://git-url/origin/project.git,https://git-url/upstream/project.git, ... }
+----
+
+When remotes are configured in this format, the remote names are implicit. 
+The first remote is `origin`, the second remote is `upstream`, and subsequent remotes are `fork1`, `fork2`, ... ,  `fork(n-2)`.
```

Comment:
so it doesn't use the /upstream/ org, but instead uses `fork1` ? Hmm. that's an odd choice ...

Reply:
@dkwon17 and @nickboldt, I'm not a developer, but I find it confusing for a few reasons:
1. I have a number of remotes in my git, and most of them are named according to their default names when they get added. For example, one of my remotes is named `dkwon17`. So as a user, I'd like to operate with whatever names those remotes get by default (that's already sufficient for me to be able to identify them in the editor of the workspace).
2. Whatever work I am doing is normally to be pulled from or pushed to `origin`, which in this case is the main repo: `pass:c,a,q[{prod-url}]#__<git_repository_url>__`. so I'd find it confusing to have an additional remote named `origin`.
3. When I read that the `origin` remote's default branch is checked out, the question that popped up in my head is: and what about the other additional remote, which branch is checked out there?

Reply:
I started editing to create my PR to the branch of this PR, but now it's not clear to me whether we should wait with this update of the docs if there will be an iteration of this feature. Should this PR be turned into a Draft in the meantime?

Reply:
Regarding examples, I'd like to see some real-life repos, for example:
```
pass:c,a,q[{prod-url}]#https://github.com/golang/example.git?remotes={{golang,https://github.com/golang/tools.git},{godoctor,https://github.com/godoctor/godoctor.git}}
```

Reply:
1. I've created an issue for the enhancement [here](https://github.com/eclipse/che/issues/21943), but that is still possible by providing explicit remote names to the `remotes` paramater (ie: `?remotes={{dkwon17,https://github.com/dkwon17/tools.git}}`)

2. If the origin branch in the `remotes` parameter will actually replace the original one, so at the end, there is still one `origin` remote.

3. I'm not sure if I understand the question, but if another remote is to be checked out and no branch is specified, it would be the default branch for that remote too.

I can update this PR for more clarity

Reply:
@dkwon17, I can complete my review then and you can see what you want to keep from my proposed changes.

To clarify regarding

> 2. The origin branch in the remotes parameter will actually replace the original one, so at the end, there is still one origin remote.

Does `https://<che_fqdn>#<git_repository_url>` from https://www.eclipse.org/che/docs/stable/end-user-guide/starting-a-new-workspace-with-a-clone-of-a-git-repository/ work together with this parameter? If yes, then I assume that repository will also be cloned, correct? So it's only about which of these repos gets named `origin`, correct? If that's the case, maybe we should not document the URL-only syntax option for now, and only document {name,URL} syntax? WDYT, the new page gets published just without the URL-only syntax option (until the new enhancement issue is resolved)?

Reply:
Yes, `https://<che_fqdn>#<git_repository_url>` from https://www.eclipse.org/che/docs/stable/end-user-guide/starting-a-new-workspace-with-a-clone-of-a-git-repository/ work together with this parameter.

>  If yes, then I assume that repository will also be cloned, correct? 

Yes, but only if `origin` remote has not been provided in in the `remotes` parameter

When accessing this URL:
```
https://<che_fqdn>#https://github.com/eclipse/che?remotes={{origin,https://github.com/other-user/che}}
```
We get the project devfile from `https://github.com/eclipse/che` and initially set the `origin` remote to `https://github.com/eclipse/che`. But since the `remotes` parameter provides a new `origin` remote, it will replace the old `origin` remote (`https://github.com/eclipse/che`) with the new one (`https://github.com/other-user/che`). This happens before any project is cloned.

In other words, in this case the devfile comes from `https://github.com/eclipse/che`, but the `origin` remote (which is also the project that we clone) is `https://github.com/other-user/che`.

This is something I should have documented to make it more clear, I'd like to do so after your review. There are also a slight change introduced by https://github.com/eclipse-che/che-dashboard/pull/704 that I would like to add to the docs.

In the interest of making the docs minimal, I'm ok with only documenting the `{name,URL}` syntax until we prioritize https://github.com/eclipse/che/issues/21943 next sprint (Jan 25 - Feb 14). After prioritization (or after the enhancement is finished), I can provide a new PR to document it again

Reply:
For some concrete examples:

#### 1. `origin` remote not provided in `remotes` url param
```
{prod-url}#https://github.com/eclipse-che/che-docs.git?remotes={{max-cx,https://github.com/max-cx/che-docs.git},{dkwon17,https://github.com/dkwon17/example.git}}
```

Creates a workspace with these remotes configured:
```
origin  -> https://github.com/eclipse-che/che-docs.git
max-cx  -> https://github.com/max-cx/che-docs.git
dkwon17 -> https://github.com/dkwon17/che-docs.git
```

#### 2. `origin` remote provided in `remotes` url param
```
{prod-url}#https://github.com/eclipse-che/che-docs.git?remotes={{origin,https://github.com/max-cx/che-docs.git},{dkwon17,https://github.com/dkwon17/example.git}}
```

Creates a workspace with these remotes configured:
```
origin  -> https://github.com/max-cx/che-docs.git
dkwon17 -> https://github.com/dkwon17/example.git
```

In both cases, the origin/main branch will be checked out on workspace startup

Reply:
@dkwon17, regarding your two examples above, does example 2 mean that https://github.com/eclipse-che/che-docs.git will not be cloned for the workspace?

Reply:
Yes, for example 2,  `https://github.com/eclipse-che/che-docs.git` will not be cloned for the workspace, and ` https://github.com/max-cx/che-docs.git` will be cloned instead